flink-connector-debezium/src/main/java/com/alibaba/ververica/cdc/debezium/DebeziumSourceFunction.java

```diff
@@ -353,12 +363,18 @@ private void snapshotHistoryRecordsState() throws Exception {
         schemaRecordsState.clear();
 
         if (engineInstanceName != null) {
-            schemaRecordsState.add(engineInstanceName);
-
-            Collection<SchemaRecord> records = StateUtils.retrieveHistory(engineInstanceName);
-            DocumentWriter writer = DocumentWriter.defaultWriter();
-            for (SchemaRecord record : records) {
-                schemaRecordsState.add(writer.write(record.toDocument()));
+            if (StateUtils.historyHasCleanup(engineInstanceName)) {
+                throw new FlinkRuntimeException(
+                        String.format(
+                                "The schema records for engine %s has been removed, checkpoint failed.",
+                                engineInstanceName));
+            } else {
+                schemaRecordsState.add(engineInstanceName);
+                Collection<SchemaRecord> records = StateUtils.retrieveHistory(engineInstanceName);
+                DocumentWriter writer = DocumentWriter.defaultWriter();
+                for (SchemaRecord record : records) {
+                    schemaRecordsState.add(writer.write(record.toDocument()));
+                }
```

Comment:
I think this is still problematic, because the getting of status and records is not atomic, which means the status is ok but we still get a null records.